File: Mobile/lib/components/product.dart
import 'package:flutter/cupertino.dart';
import 'package:vinyl_app/components/request.dart';
import 'package:vinyl_app/components/review.dart';
import 'package:vinyl_app/components/song.dart';
import 'dart:convert';

//product class
class Product {
  final int prid;
  final String pname;
  final String artist;
  final String genre;
  final String description;
  final double price;
  final int categoryId;
  final String productImgUrl;
  final int stock;
  final int isVisible;
  final int warranty;
  //final String distributor;
  final List<Song> songs;
  //final Request request;

  Product({
    @required this.prid,
    @required this.pname,
    @required this.artist,
    @required this.genre,
    @required this.description,
    @required this.price,
    @required this.categoryId,
    @required this.productImgUrl,
    @required this.stock,
    @required this.isVisible,
    @required this.warranty,
    //@required this.distributor,
    @required this.songs,
    //@required this.request,
  });

  factory Product.fromJson(Map<String, dynamic> json, [List<dynamic> list]) {
    List<Song> songList = [];
    List<Review> comments = [];
    if(list != null) {
      for (int i = 0; i < list.length; i++) {
        songList.add(Song.fromJson(list[i]));
      }
    }
    return Product(
      prid: json['prid'],
      pname: json['pname'],
      artist: json['artist'],
      genre: json['genre'],
      description: json['description'],
      price: json['price'] + .0,
      categoryId: json['categoryId'],
      productImgUrl: json['productImgUrl'],
      stock: json['stock'],
      isVisible: json['isVisible'],
      warranty: json['warranty'],
      //distributor: json['distributor'],
      songs: songList  ,
      //request: Request.fromJson(jsonDecode(json['request'])),
    );
  }
}

File: API/api/models/product.js
const Sequelize = require("sequelize");
const sequelize = require("../../config/database.js");

module.exports = (sequelize) => {
  class Product extends Sequelize.Model {
    setInvisible() {
      this.isVisible = 0;
    }

    setVisible() {
      this.isVisible = 1;
    }

    setStock(stock) {
      this.stock = stock;
    }

    increaseStock(by) {
      this.stock += by;
      return this;
    }

    decreaseStock(by) {
      this.stock -= by;
      if (this.stock < 0) {
        this.stock = 0;
      }
      return this;
    }

    applyDiscount(pr) {
      this.discountedPrice = pr;
      return this;
    }

    removeDiscount() {
      this.discountedPrice = null;
      return this;
    }
    /**
     * Helper method for defining associations.
     * This method is not a part of Sequelize lifecycle.
     * The `models/index` file will call this method automatically.
     */
    /* static associate(models) {
      // define association here
    } */
  }
  Product.init(
    {
      prid: {
        type: Sequelize.DataTypes.INTEGER,
        primaryKey: true,
        autoIncrement: true,
      },
      pname: Sequelize.DataTypes.CHAR(100),
      artist: Sequelize.DataTypes.CHAR(50),
      genre: Sequelize.DataTypes.CHAR(50),
      description: Sequelize.DataTypes.TEXT,
      price: Sequelize.DataTypes.FLOAT(10, 2),
      discountedPrice: {
        type: Sequelize.DataTypes.FLOAT(10, 2),
        allowNull: true,
      },
      categoryId: Sequelize.DataTypes.INTEGER,
      productImgUrl: Sequelize.DataTypes.TEXT,
      stock: Sequelize.DataTypes.INTEGER,
      isVisible: Sequelize.DataTypes.TINYINT((length = 1)),
      warranty: Sequelize.DataTypes.INTEGER((length = 16)),
      distributor: Sequelize.DataTypes.CHAR(50),
    },
    {
      // options
      sequelize,
      modelName: "Product",
      tableName: "product",
      timestamps: false,
      freezeTableName: true,
      underscore: true,
    }
  );
  return Product;
};
